File: execution/sync_reviews.js
const fs = require('fs');
const path = require('path');
const https = require('https');

// --- PATH CONFIGURATION ---
const PROJECT_DIR = path.resolve(__dirname, '..');
const ENV_FILE = path.join(PROJECT_DIR, '.env');
const REVIEWS_JSON = path.join(PROJECT_DIR, 'data', 'reviews.json');

// Reads environment variables manually
function loadEnv() {
    const envVars = {};
    if (fs.existsSync(ENV_FILE)) {
        const lines = fs.readFileSync(ENV_FILE, 'utf-8').split('\n');
        lines.forEach(line => {
            line = line.trim();
            if (line && !line.startsWith('#') && line.includes('=')) {
                const idx = line.indexOf('=');
                const key = line.substring(0, idx).trim();
                const val = line.substring(idx + 1).trim();
                envVars[key] = val;
            }
        });
    }
    return envVars;
}

// Translates relative publish time bi-directionally between Polish and English
function translateRelativeTime(timeStr, targetLang = 'pl') {
    const timeLower = timeStr.toLowerCase().trim();
    
    // Check if input is in Polish
    const isPolish = timeLower.includes('temu') || timeLower.includes('chwilą') || timeLower.includes('tydzień') || timeLower.includes('miesiąc') || timeLower.includes('rok');
    
    if (isPolish) {
        if (targetLang === 'pl') {
            return timeStr; // Already in Polish
        }
        
        // Translate Polish to English
        const plToEn = {
            'rok temu': 'a year ago',
            'miesiąc temu': 'a month ago',
            'tydzień temu': 'a week ago',
            'dzień temu': 'a day ago',
            'godzinę temu': 'an hour ago',
            'minutę temu': 'a minute ago',
            'przed chwilą': 'just now'
        };
        
        if (plToEn[timeLower]) return plToEn[timeLower];
        
        try {
            const parts = timeLower.split(' ');
            if (parts.length >= 3 && !isNaN(parts[0])) {
                const num = parseInt(parts[0], 10);
                const unit = parts[1];
                let mappedUnit = 'ago';
                if (unit.includes('tygod')) mappedUnit = 'weeks ago';
                else if (unit.includes('miesiąc') || unit.includes('miesiec')) mappedUnit = 'months ago';
                else if (unit.includes('dni') || unit.includes('dnia')) mappedUnit = 'days ago';
                else if (unit.includes('lat') || unit.includes('lata')) mappedUnit = 'years ago';
                else if (unit.includes('godzin')) mappedUnit = 'hours ago';
                else if (unit.includes('minut')) mappedUnit = 'minutes ago';
                
                return `${num} ${mappedUnit}`;
            }
        } catch (e) {}
        return timeStr;
    } else {
        // Input is in English (or fallback)
        if (targetLang === 'en') {
            return timeStr; // Already in English
        }
        
        // Translate English to Polish
        const enToPl = {
            'a year ago': 'rok temu',
            'years ago': 'lata temu',
            'a month ago': 'miesiąc temu',
            'months ago': 'miesiące temu',
            'a week ago': 'tydzień temu',
            'weeks ago': 'tygodnie temu',
            'a day ago': 'dzień temu',
            'days ago': 'dni temu',
            'an hour ago': 'godzinę temu',
            'hours ago': 'godziny temu',
            'a minute ago': 'minutę temu',
            'minutes ago': 'minuty temu',
            'just now': 'przed chwilą'
        };
        
        if (enToPl[timeLower]) return enToPl[timeLower];
        
        try {
            const parts = timeLower.split(' ');
            if (parts.length >= 3 && !isNaN(parts[0])) {
                const num = parseInt(parts[0], 10);
                const unit = parts[1];
                let mappedUnit = 'temu';
                if (unit.includes('year')) mappedUnit = (num >= 5 || num === 0) ? 'lat temu' : 'lata temu';
                else if (unit.includes('month')) mappedUnit = (num >= 5 || num === 0) ? 'miesięcy temu' : 'miesiące temu';
                else if (unit.includes('week')) mappedUnit = (num >= 5 || num === 0) ? 'tygodni temu' : 'tygodnie temu';
                else if (unit.includes('day')) mappedUnit = 'dni temu';
                else if (unit.includes('hour')) mappedUnit = (num >= 5 || num === 0) ? 'godzin temu' : 'godziny temu';
                else if (unit.includes('minute')) mappedUnit = (num >= 5 || num === 0) ? 'minut temu' : 'minuty temu';
                
                return `${num} ${mappedUnit}`;
            }
        } catch (e) {}
        return timeStr;
    }
}

function main() {
    console.log('[Sync] Initializing Google Reviews Node synchronizer...');
    const env = loadEnv();

    const apiKey = env.GOOGLE_PLACES_API_KEY;
    const placeId = env.GOOGLE_PLACE_ID;
    const mapsUrl = env.GOOGLE_MAPS_URL;
    const writeReviewUrl = env.GOOGLE_WRITE_REVIEW_URL;

    // Load existing reviews data as baseline
    const existingData = {
        rating: 5.0,
        reviews_count: 12,
        google_maps_url: mapsUrl || "https://maps.google.com/?cid=18151900155099238804",
        write_review_url: writeReviewUrl || "https://search.google.com/local/writereview?placeid=ChIJ-Tq9m4HPG0cRpEvz86lR4fw",
        reviews: []
    };

    if (fs.existsSync(REVIEWS_JSON)) {
        try {
            const loaded = JSON.parse(fs.readFileSync(REVIEWS_JSON, 'utf-8'));
            if (loaded && typeof loaded === 'object') {
                Object.assign(existingData, loaded);
                console.log('[Sync] Existing reviews data loaded successfully.');
            }
        } catch (e) {
            console.log(`[Sync] Warning: Could not parse existing reviews.json: ${e.message}`);
        }
    }

    // Always update links from .env
    if (mapsUrl) existingData.google_maps_url = mapsUrl;
    if (writeReviewUrl) existingData.write_review_url = writeReviewUrl;

    if (apiKey && placeId && apiKey !== 'YOUR_GOOGLE_CLOUD_API_KEY' && placeId !== 'YOUR_GOOGLE_PLACE_ID') {
        console.log(`[Sync] Querying Google Places API (New) for Place ID: ${placeId}...`);

        const postOptions = {
            hostname: 'places.googleapis.com',
            port: 443,
            path: `/v1/places/${placeId}`,
            method: 'GET',
            headers: {
                'X-Goog-Api-Key': apiKey,
                'X-Goog-FieldMask': 'rating,userRatingCount,reviews',
                'Accept': 'application/json',
                'Content-Type': 'application/json',
                'Accept-Language': 'pl'
            }
        };

        const req = https.request(postOptions, (res) => {
            let body = '';
            res.on('data', chunk => body += chunk);
            res.on('end', () => {
                try {
                    const data = JSON.parse(body);
                    if (data.rating) existingData.rating = parseFloat(data.rating);
                    if (data.userRatingCount) existingData.reviews_count = parseInt(data.userRatingCount, 10);

                    console.log(`[Sync] Success! Rating updated to ${existingData.rating} (${existingData.reviews_count} reviews)`);

                    const googleReviews = data.reviews || [];
                    if (googleReviews.length > 0) {
                        const updatedReviews = [];

                        googleReviews.forEach(gr => {
                            const author = gr.authorAttribution ? gr.authorAttribution.displayName : 'Klient';
                            const rating = parseInt(gr.rating || 5, 10);
                            const textVal = gr.text ? gr.text.text : '';
                            const avatarImg = gr.authorAttribution ? gr.authorAttribution.photoUri : '';
                            const relTime = gr.relativePublishTimeDescription || 'niedawno';

                            const datePl = translateRelativeTime(relTime, 'pl');
                            const dateEn = translateRelativeTime(relTime, 'en');

                            // Find match in current database to protect hand-curated translations
                            const matched = existingData.reviews.find(r => r.author === author);

                            if (matched) {
                                matched.rating = rating;
                                matched.date = { pl: datePl, en: dateEn };
                                if (avatarImg) matched.avatar_img = avatarImg;
                                updatedReviews.push(matched);
                                console.log(`[Sync] Merged/Updated curated review from: ${author}`);
                            } else {
                                const newRev = {
                                    author: author,
                                    rating: rating,
                                    text: {
                                        pl: textVal,
                                        en: textVal
                                    },
                                    avatar: author ? author[0] : 'U',
                                    date: {
                                        pl: datePl,
                                        en: dateEn
                                    }
                                };
                                if (avatarImg) newRev.avatar_img = avatarImg;
                                updatedReviews.push(newRev);
                                console.log(`[Sync] Added new review from Google: ${author}`);
                            }
                        });

                        // Append other curated reviews that weren't in the Google API's latest feed
                        existingData.reviews.forEach(r => {
                            if (!updatedReviews.find(ur => ur.author === r.author) && updatedReviews.length < 6) {
                                updatedReviews.push(r);
                            }
                        });

                        existingData.reviews = updatedReviews;
                    }

                    writeBack(existingData);

                } catch (err) {
                    console.error('[Sync] Error parsing Google response:', err.message);
                    console.log('Raw response was:', body);
                    writeBack(existingData); // fallback write
                }
            });
        });

        req.on('error', (e) => {
            console.error(`[Sync] API HTTP request error: ${e.message}`);
            writeBack(existingData);
        });

        req.end();

    } else {
        console.log('[Sync] No Place ID or API Key configured in .env. Skipping HTTP request.');
        writeBack(existingData);
    }
}

function writeBack(data) {
    try {
        const dataDir = path.dirname(REVIEWS_JSON);
        if (!fs.existsSync(dataDir)) {
            fs.mkdirSync(dataDir, { recursive: true });
        }
        // Save reviews.json
        fs.writeFileSync(REVIEWS_JSON, JSON.stringify(data, null, 2), 'utf-8');
        
        // Save reviews.js
        const REVIEWS_JS = path.join(dataDir, 'reviews.js');
        fs.writeFileSync(REVIEWS_JS, `window.reviewsData = ${JSON.stringify(data, null, 2)};`, 'utf-8');
        
        console.log(`[Sync] Complete! Written to reviews.json and reviews.js`);
    } catch (e) {
        console.error('[Sync] Error writing reviews files:', e.message);
    }
}

main();

File: execution/sync_reviews.py
import os
import json
import urllib.request
import urllib.error

# --- PATH CONFIGURATION ---
PROJECT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
ENV_FILE = os.path.join(PROJECT_DIR, '.env')
REVIEWS_JSON = os.path.join(PROJECT_DIR, 'data', 'reviews.json')

def load_env():
    """Reads environment variables from .env file manually to avoid external dependencies."""
    env_vars = {}
    if os.path.exists(ENV_FILE):
        with open(ENV_FILE, 'r', encoding='utf-8') as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith('#') and '=' in line:
                    key, val = line.split('=', 1)
                    env_vars[key.strip()] = val.strip()
    return env_vars

def translate_relative_time(time_str, target_lang='pl'):
    time_str_lower = time_str.lower().strip()
    
    # Check if input is in Polish
    is_polish = 'temu' in time_str_lower or 'chwilą' in time_str_lower or 'tydzień' in time_str_lower or 'miesiąc' in time_str_lower or 'rok' in time_str_lower
    
    if is_polish:
        if target_lang == 'pl':
            return time_str
            
        pl_to_en = {
            'rok temu': 'a year ago',
            'miesiąc temu': 'a month ago',
            'tydzień temu': 'a week ago',
            'dzień temu': 'a day ago',
            'godzinę temu': 'an hour ago',
            'minutę temu': 'a minute ago',
            'przed chwilą': 'just now'
        }
        
        if time_str_lower in pl_to_en:
            return pl_to_en[time_str_lower]
            
        try:
            parts = time_str_lower.split()
            if len(parts) >= 3 and parts[0].isdigit():
                num = parts[0]
                unit = parts[1]
                mapped_unit = 'ago'
                if 'tygod' in unit: mapped_unit = 'weeks ago'
                elif 'miesiąc' in unit or 'miesiec' in unit: mapped_unit = 'months ago'
                elif 'dni' in unit or 'dnia' in unit: mapped_unit = 'days ago'
                elif 'lat' in unit or 'lata' in unit: mapped_unit = 'years ago'
                elif 'godzin' in unit: mapped_unit = 'hours ago'
                elif 'minut' in unit: mapped_unit = 'minutes ago'
                
                return f"{num} {mapped_unit}"
        except Exception:
            pass
        return time_str
    else:
        # Input is in English
        if target_lang == 'en':
            return time_str
            
        en_to_pl = {
            'a year ago': 'rok temu',
            'years ago': 'lata temu',
            'a month ago': 'miesiąc temu',
            'months ago': 'miesiące temu',
            'a week ago': 'tydzień temu',
            'weeks ago': 'tygodnie temu',
            'a day ago': 'dzień temu',
            'days ago': 'dni temu',
            'an hour ago': 'godzinę temu',
            'hours ago': 'godziny temu',
            'a minute ago': 'minutę temu',
            'minutes ago': 'minuty temu',
            'just now': 'przed chwilą'
        }
        
        if time_str_lower in en_to_pl:
            return en_to_pl[time_str_lower]
            
        try:
            parts = time_str_lower.split()
            if len(parts) >= 3 and parts[0].isdigit():
                num = parts[0]
                unit = parts[1]
                
                if 'year' in unit:
                    mapped_unit = 'lat temu' if int(num) >= 5 or int(num) == 0 else 'lata temu'
                elif 'month' in unit:
                    mapped_unit = 'miesięcy temu' if int(num) >= 5 or int(num) == 0 else 'miesiące temu'
                elif 'week' in unit:
                    mapped_unit = 'tygodni temu' if int(num) >= 5 or int(num) == 0 else 'tygodnie temu'
                elif 'day' in unit:
                    mapped_unit = 'dni temu'
                elif 'hour' in unit:
                    mapped_unit = 'godzin temu' if int(num) >= 5 or int(num) == 0 else 'godziny temu'
                elif 'minute' in unit:
                    mapped_unit = 'minut temu' if int(num) >= 5 or int(num) == 0 else 'minuty temu'
                else:
                    mapped_unit = 'temu'
                    
                return f"{num} {mapped_unit}"
        except Exception:
            pass
        return time_str

def main():
    print("[Sync] Initializing Google Reviews synchronizer...")
    env_vars = load_env()
    
    api_key = env_vars.get('GOOGLE_PLACES_API_KEY')
    place_id = env_vars.get('GOOGLE_PLACE_ID')
    maps_url = env_vars.get('GOOGLE_MAPS_URL')
    write_review_url = env_vars.get('GOOGLE_WRITE_REVIEW_URL')
    
    # Load existing reviews if they exist to protect manual curation/translations
    existing_data = {
        "rating": 5.0,
        "reviews_count": 12,
        "google_maps_url": maps_url or "https://maps.google.com/?cid=18151900155099238804",
        "write_review_url": write_review_url or "https://search.google.com/local/writereview?placeid=ChIJ-Tq9m4HPG0cRpEvz86lR4fw",
        "reviews": []
    }
    
    if os.path.exists(REVIEWS_JSON):
        try:
            with open(REVIEWS_JSON, 'r', encoding='utf-8') as f:
                loaded = json.load(f)
                if isinstance(loaded, dict):
                    existing_data.update(loaded)
                    print("[Sync] Existing reviews data loaded successfully.")
        except Exception as e:
            print(f"[Sync] Warning: Could not parse existing reviews.json: {e}")
            
    # Always update profile URLs from .env if provided
    if maps_url:
        existing_data["google_maps_url"] = maps_url
    if write_review_url:
        existing_data["write_review_url"] = write_review_url

    # Check if we can do an automated Google Places API sync
    if api_key and place_id and api_key != "YOUR_GOOGLE_CLOUD_API_KEY" and place_id != "YOUR_GOOGLE_PLACE_ID":
        print(f"[Sync] Attempting live Google Places API sync for Place ID: {place_id}...")
        
        # Google Places API (New) Details URL
        url = f"https://places.googleapis.com/v1/places/{place_id}"
        
        # Fields to request
        fields = "rating,userRatingCount,reviews"
        
        # Headers
        headers = {
            "X-Goog-Api-Key": api_key,
            "X-Goog-FieldMask": fields,
            "Accept": "application/json",
            "Content-Type": "application/json",
            "Accept-Language": "pl"
        }
        
        req = urllib.request.Request(url, headers=headers)
        
        try:
            with urllib.request.urlopen(req, timeout=10) as response:
                res_data = json.loads(response.read().decode('utf-8'))
                
                # Fetch rating and review counts
                if 'rating' in res_data:
                    existing_data["rating"] = float(res_data['rating'])
                if 'userRatingCount' in res_data:
                    existing_data["reviews_count"] = int(res_data['userRatingCount'])
                    
                print(f"[Sync] Success! Rating updated to {existing_data['rating']} ({existing_data['reviews_count']} reviews)")
                
                # Format and merge reviews if available
                google_reviews = res_data.get('reviews', [])
                if google_reviews:
                    updated_reviews = []
                    
                    for gr in google_reviews:
                        author = gr.get('authorAttribution', {}).get('displayName', 'Klient')
                        rating = int(gr.get('rating', 5))
                        text_val = gr.get('text', {}).get('text', '')
                        avatar_img = gr.get('authorAttribution', {}).get('photoUri', '')
                        rel_time = gr.get('relativePublishTimeDescription', 'niedawno')
                        
                        # Translate dates
                        date_pl = translate_relative_time(rel_time, 'pl')
                        date_en = translate_relative_time(rel_time, 'en')
                        
                        # Look for existing review by the same author to protect manually curated translations
                        matched_existing = None
                        for r in existing_data.get('reviews', []):
                            if r.get('author') == author:
                                matched_existing = r
                                break
                                
                        if matched_existing:
                            # Update details but preserve manually curated/translated texts if available
                            matched_existing['rating'] = rating
                            matched_existing['date'] = {
                                "pl": date_pl,
                                "en": date_en
                            }
                            if avatar_img:
                                matched_existing['avatar_img'] = avatar_img
                            updated_reviews.append(matched_existing)
                            print(f"[Sync] Merged/Updated existing review from: {author}")
                        else:
                            # Create new review
                            new_rev = {
                                "author": author,
                                "rating": rating,
                                "text": {
                                    "pl": text_val,
                                    "en": text_val # Fallback both to same text
                                },
                                "avatar": author[0] if author else "U",
                                "date": {
                                    "pl": date_pl,
                                    "en": date_en
                                }
                            }
                            if avatar_img:
                                new_rev['avatar_img'] = avatar_img
                            updated_reviews.append(new_rev)
                            print(f"[Sync] Added new review from Google: {author}")
                            
                    # Fill remainder with existing curated reviews that weren't in the API's latest 5
                    for r in existing_data.get('reviews', []):
                        if r not in updated_reviews and len(updated_reviews) < 6:
                            updated_reviews.append(r)
                            
                    existing_data["reviews"] = updated_reviews
                    
        except urllib.error.URLError as e:
            print(f"[Sync] API connection error: {e}. Falling back to curated/local reviews.")
        except Exception as e:
            print(f"[Sync] Unexpected error during sync: {e}. Falling back to curated/local reviews.")
    else:
        print("[Sync] No Google Places API credentials configured in .env.")
        print("[Sync] Using manual/curated fallback and updating links from .env.")

    # Write final merged structure back to data/reviews.json
    try:
        os.makedirs(os.path.dirname(REVIEWS_JSON), exist_ok=True)
        with open(REVIEWS_JSON, 'w', encoding='utf-8') as f:
            json.dump(existing_data, f, indent=2, ensure_ascii=False)
        print(f"[Sync] Complete! Reviews written to {os.path.basename(REVIEWS_JSON)}")
    except Exception as e:
        print(f"[Sync] Error: Could not write final reviews.json: {e}")

if __name__ == "__main__":
    main()
